semgrep-core/src/engine/Match_rules.ml

```diff
@@ -796,48 +796,90 @@ and filter_ranges env xs cond =
+               (* the following code is very similar to `check' below;
+                * we could maybe factorize things *)
+               let xpatterns = xpatterns_in_formula formula in
+               let res =
+                 matches_of_xpatterns env.config env.rule env.equivalences
+                   (env.file, xlang, lazy_ast_and_errors, lazy_content)
+                   xpatterns
+               in
+               let pattern_matches_per_id =
+                 group_matches_per_pattern_id res.matches
+               in
+               let env =
+                 { env with pattern_matches = pattern_matches_per_id }
+               in
+               match evaluate_formula env opt_context formula with
+               | [] -> false
+               | _ :: _ -> true
+             in
              match List.assoc_opt mvar bindings with
              | None ->
+                 (* THINK: fatal error instead? *)
                  logger#error "rule %s: metavariable-pattern: %s not found"
                    env.rule.id mvar;
                  false
              | Some mval -> (
-                 match MV.program_of_mvalue mval with
-                 | None ->
-                     logger#error
-                       "rule %s: metavariable-pattern: %s does not bound a \
-                        sub-program"
-                       env.rule.id mvar;
-                     false
-                 | Some mast -> (
-                     let mval_range = MV.range_of_mvalue mval in
-                     let lazy_ast_and_errors = lazy (mast, []) in
-                     let lazy_content =
-                       lazy (Range.content_at_range env.file mval_range)
-                     in
-                     (* this following code is very similar to `check' below;
-                      * we could maybe factorize things *)
-                     let xpatterns = xpatterns_in_formula formula in
-                     let res =
-                       matches_of_xpatterns env.config env.rule env.equivalences
-                         (env.file, env.xlang, lazy_ast_and_errors, lazy_content)
-                         xpatterns
-                     in
-                     let pattern_matches_per_id =
-                       group_matches_per_pattern_id res.matches
+                 let mval_range = MV.range_of_mvalue mval in
+                 match (opt_lang, mval) with
+                 | Some lang, MV.Text (content, _) ->
+                     (* The matched text must be interpreted according to `lang'. *)
+                     let lazy_ast_and_errors =
+                       lazy_ast_of_string lang content
                      in
-                     let env =
-                       { env with pattern_matches = pattern_matches_per_id }
+                     go_eval_formula env env.xlang formula lazy_ast_and_errors
+                       (lazy content)
+                       None
+                 | Some lang, _ ->
```

Comment:
We could probably restrict here to G.E (G.L (G.String ...)) or the stuff that spacegrep and comby returns via mval_of_string.

Reply:
I did, but note that `mval_of_string` tries to interpret the string as an integer and if possible it returns an integer literal instead of a string literal. So those cases we won't be able to match.